apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+        {
+            // Pre-condition: Must be locked
+            if (_currentBatch != null && (force || CurrentBatchIsFull(_idleClients.Count == 0, extraByteCount)))
+            {
+                // Cancel the timeout for this batch.
+                _currentBatch.TimerCts.Cancel();
+                // Queue the batch ready for sending.
+                _batchesReady.Enqueue(_currentBatch);
+                // Mark that there is no current batch.
+                _currentBatch = null;
+                // Trigger send to server.
+                Task unusedSendTask = TriggerSend();
+            }
+        }
+
+        // Pre-condition: Must be locked
+        // Allow one message that is larger than requestByteThreshold
+        private bool CurrentBatchIsFull(bool isQueuing, int extraByteCount) => isQueuing ?
+                _currentBatch.Messages.Count >= _batchMaxElementCount || (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount >= _batchMaxByteCount) :
+                _currentBatch.Messages.Count >= _batchElementCountThreshold || (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount >= _batchRequestByteThreshold);
+
+        // Not "async void" to gain standard async exeption handling
+        private async Task TriggerSend()
+        {
+            // Pre-condition: Must be locked - lock active up until first 'await'
+            // Start sending a batch if there's a batch to send, and a client to use to send it.
+            if (_batchesReady.Count > 0 && _idleClients.Count > 0)
+            {
+                // Remove client and batch from relevant queues.
+                var client = _idleClients.Dequeue();
+                var batch = _batchesReady.Dequeue();
+                _batchesInFlightCount += 1;
+                // Update flow-control counts.
+                _queueElementCount -= batch.Messages.Count;
+                _queueByteCount -= batch.ByteCount;
+                // Perform the RPC to server, catching exceptions.
+                try
+                {
+                    var response = await _taskHelper.ConfigureAwait(client.PublishAsync(TopicName, batch.Messages, CallSettings.FromCancellationToken(_hardStopCts.Token)));
+                    batch.BatchCompletion.SetResult(response.MessageIds);
+                }
+                catch (Exception e)
+                {
+                    batch.BatchCompletion.SetException(e);
+                }
+                // A client is now idle. Record it, and see if a further batch is ready to send.
+                lock (_lock)
+                {
+                    _batchesInFlightCount -= 1;
+                    _idleClients.Enqueue(client);
+                    ShutdownIfCompleted();
+                    if (_batchesReady.Count > 0)
+                    {
+                        // Already a batch in the ready-queue, so just send it.
+                        Task unusedSendTask = TriggerSend();
```

Comment:
The recursion here is slightly confusing - I worry that something will be keeping track of a logical call stack, and after blasting a lot of messages, it'll get very large. I wonder whether we might want a different approach to the threading for this. Probably best talked through on VC.